### PR TITLE
#160577576 Users should be able to rate articles

### PR DESCRIPTION
**Pull request by YAHYA-H**
**What does this PR do?**
This PR lets a user rate an article, update his/her own rating and also a user is able to see
the average rating on each article, how many users rated it and  counts the number of each star rated and it displays it out as a bar graph

**Description of Task to be completed?**

A 5-star rating system each contributing to the average rating of the article
Finds out how many users rated  a particular article
Counts how many users rated each star in the rating system

**How should this be manually tested?**
```
clone/download this [repo](-https://github.com/andela/ah-codeblooded-frontend.git) to your local computer
In your terminal cd to the directory of the above repo (eg) cd Desktop/ah-codeblooded-frontent
git checkout ft-rate-160577576 
yarn install
yarn test
```

**Any background context you want to provide?**
N/A

**What are the relevant pivotal tracker stories?**

[#160577576](https://www.pivotaltracker.com/story/show/160577576)

**Screenshots (if appropriate)**

image
image

**Questions:**
N/A

### CODE REVIEW

File: src/containers/Rating/state/action.test.js [hunk 1]
```diff
@@ -0,0 +1,84 @@
+import thunk from 'redux-thunk';
+import configureMockStore from 'redux-mock-store';
+import MockAdapter from 'axios-mock-adapter';
+import api from '../../../utils/api';
+import config from '../../../utils/config';
+import {
+  rateArticle, currentRate, start, fail, success,
+} from './action';
+import { CURRENT_RATE, UPDATE_RATE, RATE_ERROR } from './types';
+
```
Comment (charisschomba) [inline src/containers/Rating/state/action.test.js:10]: have this imports in a new line

File: src/containers/Rating/state/action.test.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+      .dispatch(rateArticle(val))
+      .then(() => {
+        expect(store.getActions()).toContainEqual({
+          type: RATE_ERROR,
+          payload: err.message,
+        });
+      })
+      .catch(() => {});
+  });
+});
```
Comment (charisschomba) [inline src/containers/Rating/state/action.test.js:89]: add  have new line after every each `it` block

File: src/containers/Rating/state/reducer.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
+import { CURRENT_RATE, UPDATE_RATE, RATE_ERROR } from './types';
```
Comment (charisschomba) [inline src/containers/Rating/state/reducer.js:1]: have this imports a new line

File: src/containers/Rating/state/reducer.test.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
+import rate from './reducer';
+import { CURRENT_RATE, UPDATE_RATE, RATE_ERROR } from './types';
```
Comment (charisschomba) [inline src/containers/Rating/state/reducer.test.js:2]: same here

File: src/containers/Rating/state/reducer.test.js [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  it('should handle UPDATE_RATE', () => {
+    action.type = UPDATE_RATE;
+    action.payload = val;
+    expect(rate(initialState.state, action)).toEqual({ success: true, state: val.data.rating });
+  });
+  it('should handle RATE_ERROR', () => {
+    action.type = RATE_ERROR;
+    expect(rate(initialState.state, action)).toEqual({ failure: true, state: undefined });
+  });
+});
```
Comment (charisschomba) [inline src/containers/Rating/state/reducer.test.js:30]: add a new line after every it statement

File: src/containers/RatingStats/state/action.test.js [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+import MockAdapter from 'axios-mock-adapter';
+import api from '../../../utils/api';
+import config from '../../../utils/config';
+import {
+  fetchingRatings,
+  fetchingRatingSuccess,
+  fetchingRatingsFailure,
+  fetchRatingsAction,
+} from './actions';
+import { FETCH_RATINGS, FETCH_RATINGS_SUCCESS, FETCH_RATINGS_FAILURE } from './types';
```
Comment (charisschomba) [inline src/containers/RatingStats/state/action.test.js:12]: same here

File: src/containers/RatingStats/state/reducer.js [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+      const {
+        payload: {
+          data: { each_rating: newEachRating },
+        },
+      } = action;
+      return {
+        ...state,
+        isFetching: false,
+        rating: { ...action.payload.data, each_rating: { ...eachRating, ...newEachRating } },
+      };
```
Comment (charisschomba) [inline src/containers/RatingStats/state/reducer.js:37]: same here

File: src/containers/RatingStats/state/reducer.test.js [hunk 8]
```diff
@@ -0,0 +1,64 @@
+import ratingStats from './reducer';
+import { FETCH_RATINGS, FETCH_RATINGS_SUCCESS, FETCH_RATINGS_FAILURE } from './types';
```
Comment (charisschomba) [inline src/containers/RatingStats/state/reducer.test.js:2]: same here

**Review by charisschomba — CHANGES_REQUESTED** [review]

**Comment by YAHYA-H** [comment]
@chariss 👍

**Review by charisschomba — APPROVED** [review]

**Review by imireallan — APPROVED** [review]

File: src/containers/LikeDislike/index.js [hunk 9]
```diff
@@ -6,54 +6,51 @@ import './LikeDislike.scss';
...
-      fetch(slug);
+  componentWillMount = () => {
+    const { slug, fetch } = this.props;
+    fetch(slug);
 
-      this.interval = setInterval(() => {
-        fetch(slug);
-      }, 30000);
-    }
+    this.interval = setInterval(() => {
```
Comment (SrMoffat) [inline src/containers/LikeDislike/index.js:13]: @imireallan @YAHYA-H ,
What are we trying to achieve here with the time delay?

**Review by SrMoffat — APPROVED** [review]
@YAHYA-H ,
This is good work. I have tested it locally and the UX is very intuitive. Nice.

**Comment by MarloneA** [comment]
LGTM

**Pull request merged**